collection/collection.go

```diff
@@ -28,22 +28,217 @@ var (
+	return c.set.List(ctx, pid, lp)
+}
+
+// Get fetches dataset information as a `dsref.VersionInfo`, scoped by user
+func (c *Collection) Get(ctx context.Context, pid profile.ID, initID string) (dsref.VersionInfo, error) {
+	return c.set.Get(ctx, pid, initID)
+}
+
+// Put adds a list of `dsref.VersionInfo`s to a user's collection
+func (c *Collection) Put(ctx context.Context, pid profile.ID, item ...dsref.VersionInfo) error {
+	return c.set.Put(ctx, pid, item...)
+}
+
+func (c *Collection) subscribe(bus event.Bus) {
+	bus.SubscribeTypes(c.handleEvent,
+		// save events
+		event.ETDatasetNameInit,
+		event.ETDatasetCommitChange,
+		event.ETDatasetRename,
+		event.ETDatasetDeleteAll,
+
+		// remote & registry events
+		event.ETDatasetPushed,
+		event.ETDatasetPulled,
+		event.ETRegistryProfileCreated,
+
+		// automation events
+		event.ETAutomationWorkflowStarted,
+		event.ETAutomationWorkflowStopped,
+		event.ETAutomationWorkflowCreated,
+		event.ETAutomationWorkflowRemoved,
+
+		// fsi
+		event.ETFSICreateLink,
+		event.ETFSIRemoveLink,
+	)
+}
+
+func (c *Collection) handleEvent(ctx context.Context, e event.Event) error {
+	switch e.Type {
+	case event.ETDatasetNameInit:
+		if vi, ok := e.Payload.(dsref.VersionInfo); ok {
+			pid, err := profile.IDB58Decode(vi.ProfileID)
+			if err != nil {
+				log.Debugw("parsing profile ID in name init", "err", err)
+				return err
+			}
+			if err := c.set.Put(ctx, pid, vi); err != nil {
+				log.Debugw("putting one:", "err", err)
+				return err
+			}
+			log.Debugw("finished putting new name", "name", vi.Name, "initID", vi.InitID)
+		}
+	case event.ETDatasetCommitChange:
+		// keep in mind commit changes can mean added OR removed versions
+		if vi, ok := e.Payload.(dsref.VersionInfo); ok {
+			c.set.UpdateOneAcrossAllCollections(ctx, vi.InitID, func(m *dsref.VersionInfo) {
+				// preserve fsi path
+				vi.FSIPath = m.FSIPath
+				// preserve workflow id
+				vi.WorkflowID = m.WorkflowID
+				// preserve run information
+				vi.RunID = m.RunID
+				vi.RunStatus = m.RunStatus
+				*m = vi
+			})
+		}
+	case event.ETDatasetRename:
+		if rename, ok := e.Payload.(event.DsRename); ok {
+			c.set.UpdateOneAcrossAllCollections(ctx, rename.InitID, func(vi *dsref.VersionInfo) {
+				vi.Name = rename.NewName
+			})
+		}
+	case event.ETDatasetDeleteAll:
+		profileID := e.ProfileID
+		if initID, ok := e.Payload.(string); ok && profileID != "" {
+			if err := c.set.DeleteFromCollection(ctx, profileID, initID); err != nil {
+				log.Debugw("removing dataset from collection", "profileID", profileID, "initID", initID, "err", err)
+			}
+		}
+	case event.ETRegistryProfileCreated:
+		if p, ok := e.Payload.(event.RegistryProfileCreated); ok {
+			c.set.UpdateUsernameAcrossAllCollections(ctx, p.ProfileID, p.Username)
+		}
+	case event.ETDatasetPushed:
+		// TODO(b5): need user-scoped events for this so we can know *who* pulled
```

Comment:
I realize this is not newly written code, just moved from elsewhere, but this TODO has been on my mind. Note that this event is published on the receiver's end of a push. Since we know only local nodes use events and cloud doesn't, and currently non-cloud nodes can't push to each other's collections, I think it would be better to just log an error here and not even call `AddOneAcrossAllCollections`. Once that situation changes, it'll be easier to figure out what the handler should be doing here if it showing an error that is loud and explicit, rather than having to figure out what this incorrect code is doing.